creator-node/src/middlewares.js

```diff
@@ -389,6 +436,118 @@ async function getCreatorNodeEndpoints ({ serviceRegistry, logger, wallet, block
   return userReplicaSet
 }
 
+/**
+ * Retrieves user replica set spIDs from chain (POA.UserReplicaSetManager)
+ *
+ * Polls contract (via web3 provider) conditionally as follows:
+ *    - If `blockNumber` provided, polls contract until it has indexed that blockNumber (for up to 200 seconds)
+ *    - Else if `ensurePrimary` required, polls contract until it has indexed selfSpID as primary (for up to 60 seconds)
+ *      - Errors if retrieved primary spID does not match selfSpID
+ *    - If neither of above conditions are met, falls back to single contract query without polling
+ *
+ * @param {Object} serviceRegistry
+ * @param {Object} logger
+ * @param {string} wallet - wallet used to query discprov for user data
+ * @param {number} blockNumber - blocknumber of eth TX preceding CN call
+ * @param {string} selfSpID
+ * @param {boolean} ensurePrimary - determines if function should error if this CN is not primary
+ *
+ * @returns {Array} - array of strings of replica set
+ */
+async function getReplicaSetSpIDs ({ serviceRegistry, logger, wallet, userId, blockNumber, ensurePrimary, selfSpID }) {
+  const start = Date.now()
+  const logPrefix = `[getReplicaSetSpIDs] [userId = ${userId}]`
+  const { libs } = serviceRegistry
+
+  // returns Object of schema { primaryId: string, secondaryIds: int[] }
+  let replicaSet = null
+
+  /**
+   * If `blockNumber` provided, polls contract until it has indexed that blocknumber (for up to 200 seconds)
+   */
+  if (blockNumber) {
+    // In total, will try for 200 seconds.
+    const MaxRetries = 201
+    const RetryTimeout = 1000 // 1 seconds
+
+    let blockNumberIndexed = false
+    for (let retry = 1; retry <= MaxRetries; retry++) {
+      logger.info(`${logPrefix} retry #${retry}/${MaxRetries} || time from start: ${Date.now() - start}. Polling until blockNumber ${blockNumber}.`)
+
+      try {
+        // will throw error if blocknumber not found
+        replicaSet = await libs.contracts.UserReplicaSetManagerClient.getUserReplicaSetAtBlockNumber(userId, blockNumber)
+        blockNumberIndexed = true
+        break
+      } catch (e) { } // Ignore all errors until MaxRetries exceeded
+
```

Comment:
worth logging here?

Reply:
idts, its super noisy for no benefit - this call will always fail until it doesnt. and if it never succeeds, we already have a log for that. you see any reason to add?